File: _deliverables/indexnow-submit.py
#!/usr/bin/env python3
"""
IndexNow submitter for styermortgage.com.

Pushes URL change notifications to Bing (via IndexNow), which also feeds
AI search engines like ChatGPT search. No OAuth — auth is proven by
hosting /{KEY}.txt at the site root.

Usage:
    python3 indexnow-submit.py --from-sitemap              # submit all sitemap URLs
    python3 indexnow-submit.py --from-git-diff             # submit URLs modified since HEAD~1
    python3 indexnow-submit.py URL1 URL2 ...               # submit explicit URLs
    python3 indexnow-submit.py --from-git-diff --dry-run   # print what would be sent

Exit codes:
    0  — success (HTTP 200 or 202)
    1  — network/API error
    2  — invalid args or nothing to submit
"""
from __future__ import annotations
import sys, os, json, subprocess, argparse, urllib.request, urllib.error
from pathlib import Path

KEY = "acd320ce4aaac882bfb455892bdcf208"
HOST = "styermortgage.com"
KEY_LOCATION = f"https://{HOST}/{KEY}.txt"
ENDPOINT = "https://api.indexnow.org/IndexNow"
REPO_ROOT = Path(__file__).resolve().parent.parent

def urls_from_sitemap() -> list[str]:
    import re
    sm = (REPO_ROOT / "sitemap.xml").read_text()
    return re.findall(r"<loc>([^<]+)</loc>", sm)

def urls_from_git_diff(ref: str = "HEAD~1") -> list[str]:
    """Return URLs for HTML files changed between `ref` and working tree."""
    try:
        out = subprocess.check_output(
            ["git", "diff", "--name-only", ref, "--", "*.html"],
            cwd=REPO_ROOT, text=True,
        )
    except subprocess.CalledProcessError:
        return []
    urls = []
    for line in out.splitlines():
        line = line.strip()
        if not line or not line.endswith(".html"): continue
        # Skip pages not meant for public indexing
        if line in {"dashboard.html", "loanos.html", "hero-test.html"}: continue
        if line.startswith("updates/") or line.startswith("netlify/"): continue
        # Root HTML → /page.html; subdirs preserved
        urls.append(f"https://{HOST}/{line}")
    return urls

def submit(urls: list[str], dry_run: bool = False) -> int:
    urls = sorted(set(urls))
    if not urls:
        print("No URLs to submit.", file=sys.stderr)
        return 2
    # IndexNow accepts up to 10,000 per request; we're always well under
    payload = {
        "host": HOST,
        "key": KEY,
        "keyLocation": KEY_LOCATION,
        "urlList": urls,
    }
    print(f"Submitting {len(urls)} URL(s) to IndexNow:")
    for u in urls: print(f"  {u}")
    if dry_run:
        print("\n[--dry-run] Not sending.")
        return 0
    data = json.dumps(payload).encode("utf-8")
    req = urllib.request.Request(
        ENDPOINT, data=data, method="POST",
        headers={"Content-Type": "application/json; charset=utf-8"},
    )
    try:
        with urllib.request.urlopen(req, timeout=15) as resp:
            code = resp.getcode()
            body = resp.read().decode("utf-8", errors="replace")
    except urllib.error.HTTPError as e:
        code = e.code
        body = e.read().decode("utf-8", errors="replace")
    except Exception as e:
        print(f"Network error: {e}", file=sys.stderr)
        return 1
    print(f"\nIndexNow response: HTTP {code}")
    if body.strip(): print(body)
    # 200 = accepted, 202 = accepted (async processing).
    # 400 = bad request, 403 = key not valid, 422 = URLs don't belong to host, 429 = rate-limited.
    return 0 if code in (200, 202) else 1

def main():
    p = argparse.ArgumentParser(description=__doc__)
    p.add_argument("--from-sitemap", action="store_true", help="Submit every URL in sitemap.xml")
    p.add_argument("--from-git-diff", action="store_true", help="Submit HTML files modified since HEAD~1")
    p.add_argument("--since", default="HEAD~1", help="Git ref to diff against (default HEAD~1)")
    p.add_argument("--dry-run", action="store_true", help="Print URLs without sending")
    p.add_argument("urls", nargs="*", help="Explicit URLs to submit")
    args = p.parse_args()

    urls: list[str] = list(args.urls)
    if args.from_sitemap: urls += urls_from_sitemap()
    if args.from_git_diff: urls += urls_from_git_diff(args.since)

    if not urls:
        print("Nothing to submit. Pass URLs or --from-sitemap / --from-git-diff.", file=sys.stderr)
        return 2
    return submit(urls, dry_run=args.dry_run)

if __name__ == "__main__":
    sys.exit(main())

File: scripts/indexnow-submit.mjs
// IndexNow: tell Bing (and the engines that share its index — ChatGPT search,
// Copilot, Yandex, Naver, Seznam) which public pages changed, so they refetch
// within hours instead of weeks. Google ignores IndexNow; Search Console covers it.
//
// Runs at the end of `npm run build` on Netlify (never fails the build) and by
// hand: `npm run indexnow -- /page.html /other.html` or `npm run indexnow -- --all`.
// The key is self-issued and hosted at https://styermortgage.com/<key>.txt.
import fs from 'node:fs';
import { execFileSync } from 'node:child_process';

const HOST = 'styermortgage.com';
const KEY = fs.readdirSync('.').find(f => /^[a-f0-9]{32}\.txt$/.test(f))?.replace('.txt', '');
if (!KEY) { console.log('IndexNow: no key file found, skipping.'); process.exit(0); }

const args = process.argv.slice(2);
const sitemap = fs.readFileSync('sitemap.xml', 'utf8');
const inSitemap = new Set([...sitemap.matchAll(/<loc>([^<]+)<\/loc>/g)].map(m => m[1]));
const toUrl = file => {
  let p = '/' + file.replace(/\\/g, '/');
  if (p.endsWith('/index.html')) p = p.slice(0, -'index.html'.length);
  return `https://${HOST}${p}`;
};

let urls = [];
if (args.includes('--all')) {
  urls = [...inSitemap];
} else if (args.length) {
  urls = args.map(a => a.startsWith('http') ? a : `https://${HOST}${a.startsWith('/') ? a : '/' + a}`);
} else {
  // Netlify exposes the previously built commit; diff against it. Locally, diff HEAD~1.
  const from = process.env.CACHED_COMMIT_REF, to = process.env.COMMIT_REF || 'HEAD';
  let changed = [];
  try {
    if (from && from !== to) {
      try { execFileSync('git', ['fetch', '--depth=50', 'origin', from], { stdio: 'ignore' }); } catch {}
      changed = execFileSync('git', ['diff', '--name-only', from, to], { encoding: 'utf8' }).split('\n');
    } else if (!from) {
      changed = execFileSync('git', ['diff', '--name-only', 'HEAD~1', 'HEAD'], { encoding: 'utf8' }).split('\n');
    }
  } catch (e) { console.log('IndexNow: could not read git diff:', e.message); }
  urls = changed.filter(f => /\.html$/.test(f)).map(toUrl).filter(u => inSitemap.has(u));
}
urls = [...new Set(urls)];
if (!urls.length) { console.log('IndexNow: no public page changes to submit.'); process.exit(0); }
if (process.env.CONTEXT && process.env.CONTEXT !== 'production') { console.log(`IndexNow: ${process.env.CONTEXT} build, not submitting ${urls.length} URL(s).`); process.exit(0); }

const body = { host: HOST, key: KEY, keyLocation: `https://${HOST}/${KEY}.txt`, urlList: urls.slice(0, 10000) };
try {
  const res = await fetch('https://api.indexnow.org/indexnow', { method: 'POST', headers: { 'content-type': 'application/json; charset=utf-8' }, body: JSON.stringify(body) });
  console.log(`IndexNow: submitted ${urls.length} URL(s) — HTTP ${res.status}${res.status === 200 || res.status === 202 ? ' (accepted)' : ' ' + (await res.text()).slice(0, 200)}`);
  for (const u of urls.slice(0, 20)) console.log('  ' + u);
} catch (e) { console.log('IndexNow: submission failed (build continues):', e.message); }
